File: domain/src/main/java/com/beat/domain/staff/domain/Staff.java
package com.beat.domain.staff.domain;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Staff {

	private Long id;

	private String staffName;

	private String staffRole;

	private String staffPhoto;

	private Long performanceId;

	@Builder(access = AccessLevel.PRIVATE)
	private Staff(Long id, String staffName, String staffRole, String staffPhoto, Long performanceId) {
		this.id = id;
		this.staffName = staffName;
		this.staffRole = staffRole;
		this.staffPhoto = staffPhoto;
		this.performanceId = performanceId;
	}

	public static Staff create(String staffName, String staffRole, String staffPhoto, Long performanceId) {
		return Staff.builder()
			.staffName(staffName)
			.staffRole(staffRole)
			.staffPhoto(staffPhoto)
			.performanceId(performanceId)
			.build();
	}

	public static Staff rehydrate(Long id, String staffName, String staffRole, String staffPhoto, Long performanceId) {
		return Staff.builder()
			.id(id)
			.staffName(staffName)
			.staffRole(staffRole)
			.staffPhoto(staffPhoto)
			.performanceId(performanceId)
			.build();
	}

	public void update(String staffName, String staffRole, String staffPhoto) {
		this.staffName = staffName;
		this.staffRole = staffRole;
		this.staffPhoto = staffPhoto;
	}
}

File: domain/src/main/kotlin/com/beat/domain/staff/domain/Staff.kt
package com.beat.domain.staff.domain

import com.beat.domain.performance.domain.Performance

@ConsistentCopyVisibility
data class Staff private constructor(
    private val staffId: Id?,
    val staffName: String,
    val staffRole: String,
    val staffPhoto: String,
    private val linkedPerformanceId: Performance.Id,
) {
    fun getId(): Long? = staffId?.value

    fun getPerformanceId(): Long = linkedPerformanceId.value

    fun update(staffName: String, staffRole: String, staffPhoto: String): Staff = copy(
        staffName = staffName,
        staffRole = staffRole,
        staffPhoto = staffPhoto
    )

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            @JvmStatic
            fun from(value: Long): Id = Id(value)

            @JvmStatic
            fun fromNullable(value: Long?): Id? = value?.let(::from)
        }
    }

    companion object {
        @JvmStatic
        fun create(staffName: String, staffRole: String, staffPhoto: String, performanceId: Long): Staff = Staff(
            staffId = null,
            staffName = staffName,
            staffRole = staffRole,
            staffPhoto = staffPhoto,
            linkedPerformanceId = Performance.Id.from(performanceId)
        )

        @JvmStatic
        fun rehydrate(id: Long?, staffName: String, staffRole: String, staffPhoto: String, performanceId: Long): Staff =
            Staff(
                staffId = Id.fromNullable(id),
                staffName = staffName,
                staffRole = staffRole,
                staffPhoto = staffPhoto,
                linkedPerformanceId = Performance.Id.from(performanceId)
            )
    }
}
